### ISSUE TITLE
New rule proposal: use 'var' whenever possible (consistency, maintainability)

### ISSUE THREAD

**Issue by heaths** [DotNetAnalyzers/StyleCopAnalyzers#2193]
We use `var` in our codebase whenever possible, but as new people come on board they may be used to the tediousness of declaring local variable types. Apart from reducing the number of namespaces imported (and, therefore, collisions), it also avoid having to know exactly what type a method or property will return.

But I'd like to be able to flag local declarations whenever `var` is possible rather than try to catch it in PR reviews or later after the fact.

For example, these are examples that should use `var` when such a rule is enabled.

``` c#
ConsoleColor color = ConsoleColor.Red;
IEnumerable<string> ids = packages.Select(package => package.Id);
```

There are, of course, times when it's not possible but, in my experience, very few.

``` c#
ConsoleColor color;
Enum.TryParse(typeof(ConsoleColor), "red", out color);

string id = null;
foreach (var package in packages)
{
    if (CheckSomething(package))
    {
        id = package.Id;
    }
}
```

**Comment by sharwell** [comment]
I'm almost certain a duplicate of this was created already, but now I can't find it. There are two reasons why I don't like this rule:

1. **Readability:** StyleCop tends to prefer explicit code to implicit code. For example, omitting `private` on a method is not allowed. For consistency with StyleCop itself, I could only ever support a version of this rule which falls into two categories:
  a. Only use `var` where the type is obvious - specifically in cases where the value is a cast, `as`, or `new` expression, is not a primitive type or `string`
  b. Never use `var` except where required (certain anonymous types)
2. **Versioning:** I believe this would be the first code fix introducing C# 6+ syntax, which could cause a problem for some users. Eventually this will just not be a problem anymore, but for now it weighs against what I see as an unfavorable rule.

That said, I know the [DotNetAnalyzers/VarUsageAnalyzer](https://github.com/DotNetAnalyzers/VarUsageAnalyzer) project aims to provide this functionality, along with code-cracker and possibly new features in Visual Studio 2015 itself.

**Comment by OzieGamma** [comment]
This is included in VS15 preview.
https://www.visualstudio.com/en-us/news/releasenotes/vs15-relnotes

**Comment by heaths** [comment]
This is possible using _.editorconfig_ in VS2017. Closing.
